elastic/apm-server
# sampling: prevent duplicate publication of remotely tail-sampled trace events

## Motivation/summary

Delete events from local storage when receiving a remote sampling decision. This prevents events from being published multiple times, e.g. due to the server being restarted and observing remote sampling decisions again.

Note that we do not need to do this for local decisions. Local decisions are based on ephemeral, in-memory sampling reservoirs. Restarting the server would have the effect of discarding those reservoirs.

We should consider persisting the subscriber position, to avoid reprocessing sampled trace IDs. I've opened https://github.com/elastic/apm-server/issues/4437 for that.

## Checklist

~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

I have considered changes for:
~- [ ] documentation~
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

1. Set up two APM Servers with tail-sampling enabled
1. Send a distributed trace with the root sent to server 1, and downstream transaction sent to server 2
1. Wait for both servers to publish events of a trace
1. Restart server 2
1. Wait ~30 seconds, ensure no duplicate events are published (e.g. check the distributed trace)

## Related issues

Closes #4359
